Failure in rc/test_ucp_tag_mem_type.xfer_mismatch_length/55

### Describe the bug
Happened here: https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/6805/logs/155 on swx-rdmz-ucx-gpu-02

```
2020-06-09T01:19:15.1290779Z [ RUN      ] rc/test_ucp_tag_mem_type.xfer_mismatch_length/55 <rc_v,cuda_copy,rocm_copy>
2020-06-09T01:19:15.1806058Z [     INFO ] TEST: host <-> cuda length :9437270
2020-06-09T01:19:15.2400191Z [swx-rdmz-ucx-gpu-02:13714:0:13714]        rndv.c:506  Assertion `rndv_req->send.length >= (offset + length)' failed: send_length=43, offset=0, length=65
2020-06-09T01:19:15.6653101Z 
2020-06-09T01:19:15.6655238Z /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/tag/rndv.c: [ ucp_rndv_progress_rma_get_zcopy_inner() ]
2020-06-09T01:19:15.6656776Z       ...
2020-06-09T01:19:15.6657373Z       501 
2020-06-09T01:19:15.6658061Z       502     /* ensure that tail (rest of message) is over min_zcopy */
2020-06-09T01:19:15.6659198Z       503     ucs_assertv(rndv_req->send.length >= (offset + length),
2020-06-09T01:19:15.6660095Z ==>   504                 "send_length=%zu, offset=%zu, length=%zu",
2020-06-09T01:19:15.6661169Z       505                 rndv_req->send.length, offset, length);
2020-06-09T01:19:15.6662344Z       506     tail = rndv_req->send.length - (offset + length);
2020-06-09T01:19:15.6663246Z       507     if (ucs_unlikely((tail != 0) && (tail < min_zcopy))) {
2020-06-09T01:19:15.6663866Z 
2020-06-09T01:19:16.0843470Z ==== backtrace (tid:  13714) ====
2020-06-09T01:19:16.0845435Z  0 0x0000000000059b18 ucs_debug_print_backtrace()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucs/debug/debug.c:656
2020-06-09T01:19:16.0847378Z  1 0x0000000000055cde ucp_rndv_progress_rma_get_zcopy_inner()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/tag/rndv.c:504
2020-06-09T01:19:16.0849383Z  2 0x0000000000055cde ucp_rndv_progress_rma_get_zcopy()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/tag/rndv.c:435
2020-06-09T01:19:16.0850754Z  3 0x000000000005bc0d ucp_request_try_send()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_request.inl:213
2020-06-09T01:19:16.0852254Z  4 0x000000000005bc0d ucp_request_send()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_request.inl:248
2020-06-09T01:19:16.0853775Z  5 0x000000000005bc0d ucp_rndv_send_frag_get_mem_type()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/tag/rndv.c:765
2020-06-09T01:19:16.0855138Z  6 0x000000000005bc0d ucp_rndv_recv_start_get_pipeline()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/tag/rndv.c:823
2020-06-09T01:19:16.0856469Z  7 0x000000000005bc0d ucp_rndv_matched_inner()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/tag/rndv.c:1123
2020-06-09T01:19:16.0857744Z  8 0x000000000005bc0d ucp_rndv_matched()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/tag/rndv.c:1042
2020-06-09T01:19:16.0859051Z  9 0x000000000005dcae ucp_rndv_process_rts()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/tag/rndv.c:1160
2020-06-09T01:19:16.0860349Z 10 0x000000000005dcae ucp_rndv_process_rts()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/tag/rndv.c:1164
2020-06-09T01:19:16.0861661Z 11 0x000000000005df46 ucp_rndv_rts_handler_inner()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/tag/rndv.c:1184
2020-06-09T01:19:16.0863030Z 12 0x000000000002fff3 uct_iface_invoke_am()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/base/uct_iface.h:635
2020-06-09T01:19:16.0864429Z 13 0x000000000002fff3 uct_rc_verbs_iface_handle_am()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/ib/rc/verbs/rc_verbs_impl.h:96
2020-06-09T01:19:16.0865871Z 14 0x000000000002fff3 uct_rc_verbs_iface_poll_rx_common()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/ib/rc/verbs/rc_verbs_impl.h:125
2020-06-09T01:19:16.0867342Z 15 0x000000000002fff3 uct_rc_verbs_iface_progress()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/ib/rc/verbs/rc_verbs_iface.c:133
2020-06-09T01:19:16.0868718Z 16 0x000000000002eb6a ucs_callbackq_dispatch()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucs/datastruct/callbackq.h:211
2020-06-09T01:19:16.0870047Z 17 0x000000000002eb6a uct_worker_progress()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/api/uct.h:2342
2020-06-09T01:19:16.0871359Z 18 0x000000000002eb6a ucp_worker_progress()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_worker.c:2032
2020-06-09T01:19:16.0872772Z 19 0x0000000000865bdb ucp_test_base::entity::progress()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/ucp_test.cc:736
2020-06-09T01:19:16.0874168Z 20 0x0000000000865c87 ucp_test::progress()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/ucp_test.cc:138
2020-06-09T01:19:16.0875787Z 21 0x00000000007f2930 test_ucp_tag::wait()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_tag.cc:129
2020-06-09T01:19:16.0877237Z 22 0x00000000007e7840 test_ucp_tag_mem_type::do_xfer()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_tag_mem_type.cc:132
2020-06-09T01:19:16.0878812Z 23 0x00000000007ec39a test_ucp_tag_mem_type_xfer_mismatch_length_Test::test_body()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_tag_mem_type.cc:204
2020-06-09T01:19:16.0880484Z 24 0x00000000007ec39a test_ucp_tag_mem_type_xfer_mismatch_length_Test::test_body()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_tag_mem_type.cc:205
2020-06-09T01:19:16.0882046Z 25 0x00000000005ae926 ucs::test_base::run()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/test.cc:290
2020-06-09T01:19:16.0883448Z 26 0x00000000005ae926 ucs::test_base::TestBodyProxy()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/test.cc:316
2020-06-09T01:19:16.0885172Z 27 0x0000000000591fc9 HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:3562
2020-06-09T01:19:16.0886666Z 28 0x000000000058620d testing::Test::Run()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:3635
2020-06-09T01:19:16.0888112Z 29 0x00000000005862e5 testing::TestInfo::Run()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:3812
2020-06-09T01:19:16.0889524Z 30 0x000000000058644f testing::TestCase::Run()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:3930
2020-06-09T01:19:16.0891004Z 31 0x000000000058abfd testing::internal::UnitTestImpl::RunAllTests()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:5808
2020-06-09T01:19:16.0892867Z 32 0x000000000058aee0 testing::internal::UnitTestImpl::RunAllTests()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:5725
2020-06-09T01:19:16.0894274Z 33 0x0000000000529488 RUN_ALL_TESTS()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest.h:20059
2020-06-09T01:19:16.0895548Z 34 0x0000000000529488 main()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/main.cc:102
2020-06-09T01:19:16.0896401Z 35 0x00000000000223d5 __libc_start_main()  ???:0
2020-06-09T01:19:16.0897066Z 36 0x0000000000570571 _start()  ???:0
2020-06-09T01:19:16.0897620Z =================================
2020-06-09T01:19:16.0898496Z [swx-rdmz-ucx-gpu-02:13714:0:13714] Process frozen...
2020-06-09T03:50:01.3210842Z make: *** [test] Terminated
```

--- a/src/ucp/tag/rndv.c
+++ b/src/ucp/tag/rndv.c
@@ -50,7 +50,8 @@ static int ucp_rndv_is_recv_pipeline_needed(ucp_request_t *rndv_req,
 
 static int ucp_rndv_is_put_pipeline_needed(uintptr_t remote_address,
                                            size_t length, size_t min_get_zcopy,
-                                           size_t max_get_zcopy) {
+                                           size_t max_get_zcopy)
+{
     /* fallback to PUT pipeline if remote mem type is non-HOST memory OR
      * can't do GET ZCOPY */
     return ((remote_address == 0) || (max_get_zcopy == 0) ||
@@ -224,6 +225,52 @@ ucs_status_t ucp_tag_send_start_rndv(ucp_request_t *sreq)
     }
 
     return status;
+}
+
+static UCS_F_ALWAYS_INLINE size_t
+ucp_rndv_adjust_zcopy_length(size_t min_zcopy, size_t max_zcopy, size_t align,
+                             size_t send_length, size_t offset, size_t length)
+{
+    size_t result_length, tail;
+
+    ucs_assert(length > 0);
+
+    /* ensure that the current length is over min_zcopy */
+    result_length = ucs_max(length, min_zcopy);
+
+    /* ensure that the current length is less than max_zcopy */
+    result_length = ucs_min(result_length, max_zcopy);
+
+    /* ensure that tail (rest of message) is over min_zcopy */
+    ucs_assertv(send_length >= (offset + result_length),
+                "send_length=%zu, offset=%zu, length=%zu",
+                send_length, offset, result_length);
+    tail = send_length - (offset + result_length);
+    if (ucs_unlikely((tail != 0) && (tail < min_zcopy))) {
+        /* ok, tail is less zcopy minimal & could not be processed as
+         * standalone operation */
+        /* check if we have room to increase current part and not
+         * step over max_zcopy */
+        if (result_length < (max_zcopy - tail)) {
+            /* if we can increase length by min_zcopy - let's do it to
+             * avoid small tail (we have limitation on minimal get zcopy) */
+            result_length += tail;
+        } else {
+            /* reduce current length by align or min_zcopy value
+             * to process it on next round */
+            ucs_assert(result_length > ucs_max(min_zcopy, align));
+            result_length -= ucs_max(min_zcopy, align);
+        }
+    }
+
+    ucs_assertv(result_length >= min_zcopy, "length=%zu, min_zcopy=%zu",
+                result_length, min_zcopy);
+    ucs_assertv(((send_length - (offset + result_length)) == 0) ||
+                ((send_length - (offset + result_length)) >= min_zcopy),
+                "send_length=%zu, offset=%zu, length=%zu, min_zcopy=%zu",
+                send_length, offset, result_length, min_zcopy);
+
+    return result_length;
 }
 
 static void ucp_rndv_complete_send(ucp_request_t *sreq, ucs_status_t status)
@@ -449,7 +496,6 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_rma_get_zcopy, (self),
     uct_rkey_t uct_rkey;
     size_t min_zcopy;
     size_t max_zcopy;
-    size_t tail;
     int pending_add_res;
     ucp_lane_index_t lane;
 
@@ -490,44 +536,16 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_rma_get_zcopy, (self),
     if ((offset == 0) && (remaining > 0) && (rndv_req->send.length > ucp_mtu)) {
         length = ucp_mtu - remaining;
     } else {
-        chunk = ucs_align_up((size_t)(ucs_min(rndv_req->send.length /
-                                              rndv_req->send.rndv_get.lane_count,
-                                              max_zcopy) * config->tag.rndv.scale[lane]),
+        chunk = ucs_align_up((size_t)(rndv_req->send.length /
+                                      rndv_req->send.rndv_get.lane_count
+                                      * config->tag.rndv.scale[lane]),
                              align);
         length = ucs_min(chunk, rndv_req->send.length - offset);
     }
 
-    /* ensure that the current length is over min_zcopy */
-    length = ucs_max(length, min_zcopy);
-
-    /* ensure that tail (rest of message) is over min_zcopy */
-    ucs_assertv(rndv_req->send.length >= (offset + length),
-                "send_length=%zu, offset=%zu, length=%zu",
-                rndv_req->send.length, offset, length);
-    tail = rndv_req->send.length - (offset + length);
-    if (ucs_unlikely((tail != 0) && (tail < min_zcopy))) {
-        /* ok, tail is less get zcopy minimal & could not be processed as
-         * standalone operation */
-        /* check if we have room to increase current part and not
-         * step over max_zcopy */
-        if (length < (max_zcopy - tail)) {
-            /* if we can increase length by min_zcopy - let's do it to
-             * avoid small tail (we have limitation on minimal get zcopy) */
-            length += tail;
-        } else {
-            /* reduce current length by align or min_zcopy value
-             * to process it on next round */
-            ucs_assert(length > ucs_max(min_zcopy, align));
-            length -= ucs_max(min_zcopy, align);
-        }
-    }
-
-    ucs_assertv(length >= min_zcopy, "length=%zu, min_zcopy=%zu",
-                length, min_zcopy);
-    ucs_assertv(((rndv_req->send.length - (offset + length)) == 0) ||
-                ((rndv_req->send.length - (offset + length)) >= min_zcopy),
-                "send_length=%zu, offset=%zu, length=%zu, min_zcopy=%zu",
-                rndv_req->send.length, offset, length, min_zcopy);
+    length = ucp_rndv_adjust_zcopy_length(min_zcopy, max_zcopy, align,
+                                          rndv_req->send.length, offset,
+                                          length);
 
     ucs_trace_data("req %p: offset %zu remainder %zu rma-get to %p len %zu lane %d",
                    rndv_req, offset, remaining,
@@ -791,11 +809,17 @@ ucp_rndv_recv_start_get_pipeline(ucp_worker_h worker, ucp_request_t *rndv_req,
                                  const void *rkey_buffer, uint64_t remote_address,
                                  size_t size, size_t base_offset)
 {
-    ucp_context_h context = worker->context;
+    ucp_ep_h ep             = rndv_req->send.ep;
+    ucp_ep_config_t *config = ucp_ep_config(ep);
+    ucp_context_h context   = worker->context;
     ucs_status_t status;
     size_t max_frag_size, offset, length;
+    size_t min_zcopy, max_zcopy;
 
-    max_frag_size                          = context->config.ext.rndv_frag_size;
+    min_zcopy                              = config->tag.rndv.min_get_zcopy;
+    max_zcopy                              = config->tag.rndv.max_get_zcopy;
+    max_frag_size                          = ucs_min(context->config.ext.rndv_frag_size,
+                                                     max_zcopy);
     rndv_req->send.rndv_get.remote_request = remote_request;
     rndv_req->send.rndv_get.remote_address = remote_address - base_offset;
     rndv_req->send.rndv_get.rreq           = rreq;
@@ -806,7 +830,7 @@ ucp_rndv_recv_start_get_pipeline(ucp_worker_h worker, ucp_request_t *rndv_req,
      * Step 1: GET remote fragment into HOST fragment buffer
      * Step 2: PUT from fragment buffer to MEM TYPE destination
      * Step 3: Send ATS for RNDV request
-    */
+     */
 
     status = ucp_ep_rkey_unpack(rndv_req->send.ep, rkey_buffer,
                                 &rndv_req->send.rndv_get.rkey);
@@ -817,7 +841,8 @@ ucp_rndv_recv_start_get_pipeline(ucp_worker_h worker, ucp_request_t *rndv_req,
 
     offset = 0;
     while (offset != size) {
-        length = ucs_min(size - offset, max_frag_size);
+        length = ucp_rndv_adjust_zcopy_length(min_zcopy, max_frag_size, 0,
+                                              size, offset, size - offset);
 
         /* GET remote fragment into HOST fragment buffer */
         ucp_rndv_send_frag_get_mem_type(rndv_req, remote_request, length,
@@ -1397,14 +1422,17 @@ UCS_PROFILE_FUNC_VOID(ucp_rndv_put_pipeline_frag_get_completion, (self, status),
 static ucs_status_t ucp_rndv_send_start_put_pipeline(ucp_request_t *sreq,
                                                      ucp_rndv_rtr_hdr_t *rndv_rtr_hdr)
 {
-    ucp_worker_h worker   = sreq->send.ep->worker;
-    ucp_context_h context = worker->context;
+    ucp_ep_h ep             = sreq->send.ep;
+    ucp_ep_config_t *config = ucp_ep_config(ep);
+    ucp_worker_h worker     = sreq->send.ep->worker;
+    ucp_context_h context   = worker->context;
     const uct_md_attr_t *md_attr;
     ucp_request_t *freq;
     ucp_request_t *fsreq;
     ucp_md_index_t md_index;
     size_t max_frag_size, rndv_size, length;
     size_t offset, rndv_base_offset;
+    size_t min_zcopy, max_zcopy;
 
     ucp_trace_req(sreq, "using put rndv pipeline protocol");
 
@@ -1412,7 +1440,7 @@ static ucs_status_t ucp_rndv_send_start_put_pipeline(ucp_request_t *sreq,
      * Step 1: GET fragment from send buffer to HOST fragment buffer
      * Step 2: PUT from fragment HOST buffer to remote HOST fragment buffer
      * Step 3: send ATP for each fragment request
-    */
+     */
 
     /* check if lane supports host memory, to stage sends through host memory */
     md_attr = ucp_ep_md_attr(sreq->send.ep, sreq->send.lane);
@@ -1420,8 +1448,10 @@ static ucs_status_t ucp_rndv_send_start_put_pipeline(ucp_request_t *sreq,
         return UCS_ERR_UNSUPPORTED;
     }
 
+    min_zcopy        = config->tag.rndv.min_put_zcopy;
+    max_zcopy        = config->tag.rndv.max_put_zcopy;
     rndv_size        = ucs_min(rndv_rtr_hdr->size, sreq->send.length);
-    max_frag_size    = context->config.ext.rndv_frag_size;
+    max_frag_size    = ucs_min(context->config.ext.rndv_frag_size, max_zcopy);
     rndv_base_offset = rndv_rtr_hdr->offset;
 
     /* initialize send req state on first fragment rndv request */
@@ -1451,7 +1481,8 @@ static ucs_status_t ucp_rndv_send_start_put_pipeline(ucp_request_t *sreq,
 
     offset = 0;
     while (offset != rndv_size) {
-        length = ucs_min(rndv_size - offset, max_frag_size);
+        length = ucp_rndv_adjust_zcopy_length(min_zcopy, max_frag_size, 0,
+                                              rndv_size, offset, rndv_size - offset);
 
         if (UCP_MEM_IS_ACCESSIBLE_FROM_CPU(sreq->send.mem_type)) {
             /* sbuf is in host, directly do put */

--- a/test/gtest/ucp/test_ucp_tag_mem_type.cc
+++ b/test/gtest/ucp/test_ucp_tag_mem_type.cc
@@ -183,7 +183,7 @@ UCS_TEST_P(test_ucp_tag_mem_type, xfer_mismatch_length)
 
     UCS_TEST_MESSAGE << "TEST: "
                      << ucs_memory_type_names[m_send_mem_type] << " <-> "
-                     << ucs_memory_type_names[m_recv_mem_type] << " length :"
+                     << ucs_memory_type_names[m_recv_mem_type] << " length: "
                      << length;
 
     mem_buffer m_recv_mem_buf(length, m_recv_mem_type);

